data_structures/doubly_linked_list.js

```diff
@@ -48,37 +71,218 @@ class Node {
-    remove_head() {}
+    remove_head() {
+        if(!this.head){
+            return null
+        }
+
+        let head = this.head
+
+        if(this.head === this.tail){
+            this.head = null
+            this.tail = null
+            return head.value
+        }
+
+        this.head = head.next
+        this.head.prev = null
+        return head.value
+    }
     
     //returns the max value within the list
-    get_max(){}
+    get_max(){
+        if(!this.head){
+            return null
+        }
+        else if(this.head === this.tail){
+            return this.head.value
+        }
+
+        let maxNum = Number.MIN_VALUE
+        let curr_node = this.head;
+
+        while(curr_node !== null){
+            if(maxNum < curr_node.value){
+                maxNum = curr_node.value
+            }
+            
+            curr_node = curr_node.next
+        }
+
+        return maxNum
+    }
     
     //returns the length of the list
-    len(){}
+    len(){
+        let current_node = this.head
+        let length = 0
+
+        while(current_node !== null){
+            length += 1
+
+            current_node = current_node.next
+        }
+
+        return length
+    }
     
     //removes the input node from its current spot in the 
     // List and inserts it as the new head node of the List
-    move_to_front(){}
+    move_to_front(node){
+        let current_head = this.head
+        let node_prev = node.prev
+        let node_next = node.next
+
+        if(node === this.tail){
+            this.tail = node_prev
+            this.tail.next = node_next
+            this.head = node
+            node.next = current_head
+            current_head.prev = node
+
+            return this
+        }
+
+        node_prev.next = node_next
+        node_next.prev = node_prev
+        this.head = node
+        current_head.prev = node
+        node.prev = null
+        node.next = current_head
+
+        return this
+    }
     
     //removes the input node from its current spot in the 
     // List and inserts it as the new tail node of the List
-    move_to_end(){}
+
+    //STOPPED HERE:
+    move_to_end(node){
+        let current_tail = this.tail
+
+        if(node === this.head){
+            this.head = this.head.next
+            this.head.next = node
+            this.head.prev = null
+            this.tail = node
+            this.tail.prev = current_tail
+            node.next = null
+
+            return this
+        }
+
+        let node_prev = node.prev
+        let node_next = node.next
+
+        node_prev.next = node_next
+        node_next.prev = node_prev
+        this.tail = node
+        node.prev = current_tail
+        node.next = null
+        current_tail.next = node
+
+        return this
+    }
     
     //removes a node from the list 
     //handles cases where the node was the head or tail
-    delete(){}
+    delete(node){
+        let current_head = this.head
```

Comment:
instead of setting these to variables you could just use this.head and this.tail below